packages/server/dataloader/LocalCache.ts

```diff
@@ -0,0 +1,177 @@
+      }
+    }
+  }
+  private resolveCacheHits(cacheHits: (() => void)[]) {
+    // cacheHits are possibly from the previous batch, hence the param
+    for (let i = 0; i < cacheHits.length; i++) {
+      cacheHits[i]()
+    }
+  }
+  private dispatchBatch = async () => {
+    this.hasReadDispatched = true
+    // grab a reference to them now because after await is called they may be overwritten when a new batch is created
+    const {keys, callbacks, cacheHits} = this
+    if (keys.length === 0) {
+      this.resolveCacheHits(cacheHits)
+      return
+    }
+    try {
+      const values = await this.redisCache.read(keys)
+      this.resolveCacheHits(cacheHits)
+      for (let i = 0; i < callbacks.length; i++) {
+        const {resolve, reject} = callbacks[i]
+        const value = values[i]
+        const handle = value instanceof Error ? reject : resolve
+        handle(value)
+      }
+    } catch (e) {
+      this.resolveCacheHits(cacheHits)
+      for (let i = 0; i < callbacks.length; i++) {
+        this.clearLocal(keys[i])
+        const {reject} = callbacks[i]
+        reject(e)
+      }
+    }
+  }
+
+  private clearLocal(key: string) {
+    delete this.cacheMap[key]
+    return this
+  }
+  private primeLocal(key: string, doc: Doc) {
+    this.cacheMap[key] = {
+      ts: Date.now(),
+      promise: Promise.resolve(doc)
+    }
+  }
+
+  private dispatchWrite = async () => {
+    this.hasWriteDispatched = true
+    const {writes} = this
+    const results = await this.redisCache.write(writes)
+    writes.forEach(({resolve, table, id}, idx) => {
+      const key = `${table}:${id}`
+      const result = results[idx]
+      this.primeLocal(key, result)
+      resolve(result)
+    })
+  }
+  async clear<T extends keyof RethinkTypes>(table: T, id: string) {
+    const key = `${table}:${id}`
+    this.clearLocal(key)
+    await this.redisCache.clear(key)
+    return this
+  }
+
+  async prime<T extends keyof RethinkTypes>(table: T, docs: RType<T>[]) {
+    docs.forEach((doc) => {
+      const key = `${table}:${doc.id}`
+      this.primeLocal(key, doc)
+    })
+    this.redisCache.prime(table, docs)
+    return this
+  }
+  async read<T extends keyof RethinkTypes>(table: T, id: string) {
+    const key = `${table}:${id}`
+    if (this.hasReadDispatched) {
+      this.hasReadDispatched = false
+      this.keys = []
+      this.callbacks = []
+      this.cacheHits = []
+      resolvedPromise.then(() => {
+        process.nextTick(this.dispatchBatch)
+      })
+    }
+    const cachedRecord = this.cacheMap[key]
+    if (cachedRecord) {
+      cachedRecord.ts = Date.now()
+      return new Promise<RType<T>>((resolve) => {
+        this.cacheHits.push(() => {
+          resolve(cachedRecord.promise)
+        })
+      })
+    }
+    this.keys.push(key)
+    const promise = new Promise<RType<T>>((resolve, reject) => {
+      this.callbacks.push({resolve, reject})
+    })
+    this.cacheMap[key] = {ts: Date.now(), promise}
```

Comment:
I don't quite follow this part. It looks like whether we find the `cacheRecord` in `cacheMap` or not, we will always update its `ts` to current time?

Reply:
ah yeah, i'm going for a least frequently used (LFU) cache, so it updates the last time accessed. should probably use a better variable name like `lastAccessedAt`